Implement wxTextAttr styles in wxTextCtrl (Cocoa):

Dan-Korn opened the pull request:
Add wxTEXT_ATTR_PARAGRAPH_USER_DATA and wxTextAttr::m_paragraphUserData,
for arbitrary paragraph-level data storage (implemented in Cocoa only).
Add enum wxTextAttrLineSpacingMode, and wxTextAttr::m_lineSpacingMode,
to support both absolute (exact) and auto (multiple) leading.
Change wxTextAttr::m_fontSize to have type float, and accessor/mutator
GetFractionalFontSize() and SetFractionalFontSize().
Add double wxTextAttr::m_unitsPerPoint, and accessor/mutator, to allow
better granularity for sizing, especially since most platforms' text
renderings are still based on points rather than metric millimeters.

Add support for scaling/zoom and undo/redo in wxTextCtrl on Mac (Cocoa),
as well as default handling of Cmd-Z to invoke Undo and Cmd-Y (or
Cmd-Shift-Z) to invoke Redo.

Bulk of changes are in in wxNSTextViewControl::GetStyle() and
wxNSTextViewControl::SetStyle() in src/osx/cocoa/textctrl.mm, to add
support in the NSTextView-based wxTextCtrl in Cocoa for underline,
strikethrough, super/subscript, paragraph user data, alignment,
indents, space above and below para, leading (line spacing), right-
to-left text, hyphenation, tabs, and foreground and background colors.
Care has been taken to ensure that only flagged style attributes are
set in ranges of text, and that other attributes in the range are not
all collapsed to have the entire run with all the same attributes.

In wxTextAttr::Combine(), don't apply foreground and background colors
unless they've been explicitly set.

Dan-Korn committed:
Implement wxTextAttr styles in wxTextCtrl (Cocoa):
Add wxTEXT_ATTR_PARAGRAPH_USER_DATA and wxTextAttr::m_paragraphUserData,
for arbitrary paragraph-level data storage (implemented in Cocoa only).
Add enum wxTextAttrLineSpacingMode, and wxTextAttr::m_lineSpacingMode,
to support both absolute (exact) and auto (multiple) leading.
Change wxTextAttr::m_fontSize to have type float, and accessor/mutator
GetFractionalFontSize() and SetFractionalFontSize().
Add double wxTextAttr::m_unitsPerPoint, and accessor/mutator, to allow
better granularity for sizing, especially since most platforms' text
renderings are still based on points rather than metric millimeters.

Add support for scaling/zoom and undo/redo in wxTextCtrl on Mac (Cocoa),
as well as default handling of Cmd-Z to invoke Undo and Cmd-Y (or
Cmd-Shift-Z) to invoke Redo.

Bulk of changes are in in wxNSTextViewControl::GetStyle() and
wxNSTextViewControl::SetStyle() in src/osx/cocoa/textctrl.mm, to add
support in the NSTextView-based wxTextCtrl in Cocoa for underline,
strikethrough, super/subscript, paragraph user data, alignment,
indents, space above and below para, leading (line spacing), right-
to-left text, hyphenation, tabs, and foreground and background colors.
Care has been taken to ensure that only flagged style attributes are
set in ranges of text, and that other attributes in the range are not
all collapsed to have the entire run with all the same attributes.

In wxTextAttr::Combine(), don't apply foreground and background colors
unless they've been explicitly set.

Dan-Korn committed:
Fix build errors for 'Compiler: clang Xcode: xcode6.4 C++' and 'Compiler: clang C++'

Dan-Korn committed:
Fix more build errors for 'Compiler: clang Xcode: xcode6.4 C++'

vadz commented on src/osx/textctrl_osx.cpp:
```diff
@@ -336,7 +336,7 @@ void wxTextCtrl::OnDropFiles(wxDropFilesEvent& event)
 void wxTextCtrl::OnKeyDown(wxKeyEvent& event)
 {
-    if ( event.GetModifiers() == wxMOD_CONTROL )
+    if ( event.ControlDown() )
```
This looks wrong because it means that `Cmd+Shift+C` will now copy text, for example, while we really only want `Cmd+C` do it. I.e. this relaxes the check for the existing keys, in addition to allowing using Shift for Redo.

vadz commented on include/wx/osx/textentry.h:
```diff
@@ -72,6 +72,11 @@ class WXDLLIMPEXP_CORE wxTextEntry: public wxTextEntryBase
     virtual bool CanUndo() const wxOVERRIDE;
     virtual bool CanRedo() const wxOVERRIDE;
 
+    virtual void EmptyUndoBuffer();
```
```suggestion
    virtual void EmptyUndoBuffer() wxOVERRIDE;
```
to avoid clang warnings about inconsistent use of `override`.

vadz commented on include/wx/textctrl.h:
```diff
@@ -517,10 +539,13 @@ class WXDLLIMPEXP_CORE wxTextAttr
                         m_colBack;
-    int                 m_fontSize;
+    float               m_fontSize;
+    int                 m_PixelSize;
```
```suggestion
    int                 m_pixelSize;
```
I can do it myself when applying, but the code in wx should follow the `camelCase` naming convention (even with an extra `m_` prefix).

vadz commented on interface/wx/textctrl.h:
```diff
@@ -205,7 +205,8 @@ enum wxTextAttrEffects
     wxTEXT_ATTR_EFFECT_RTL                   = 0x00000400,
-    wxTEXT_ATTR_EFFECT_SUPPRESS_HYPHENATION  = 0x00001000
+    wxTEXT_ATTR_EFFECT_SUPPRESS_HYPHENATION  = 0x00001000,
+    wxTEXT_ATTR_EFFECT_KERNING               = 0x00002000,
```
```suggestion
    wxTEXT_ATTR_EFFECT_KERNING               = 0x00002000, ///< Available since wxWidgets 3.1.3.
```

vadz commented on src/common/textcmn.cpp:
```diff
@@ -453,14 +476,15 @@ wxFont wxTextAttr::GetFont() const
     else
     {
         wxFont font(fontSize, fontFamily, fontStyle, fontWeight, underlined, fontFaceName, encoding);
+        font.SetFractionalPointSize(fontSize);
```
It would be better to construct `wxFont` from `wxFontInfo`, which allows to specify the fractional font size directly, instead of creating it with the int font size and then setting it again.

Dan-Korn replied:
Just following the example a few lines up, at line 479, which was this before my changes.

wxFont font(wxSize(0, fontSize), fontFamily, fontStyle, fontWeight, underlined, fontFaceName, encoding);

Dan-Korn replied:
Besides, this hardly seems cleaner:

        wxFont font(wxFontInfo(fontSize).Family(fontFamily).Weight(wxFontBase::GetNumericWeightOf(fontWeight)).Underlined(underlined).FaceName(fontFaceName).Encoding(encoding));

vadz commented on src/osx/cocoa/textctrl.mm:
```diff
@@ -771,6 +829,12 @@ - (void)controlTextDidEndEditing:(NSNotification *)aNotification
     [tv setDelegate: tv];
 
+    m_undoManager = nil;
+    if ( wxPeer && wxPeer->GetParent() && wxPeer->GetParent()->GetPeer() && wxPeer->GetParent()->GetPeer()->GetWXWidget() )
```
Pretty minor, but this would read a lot better with some extra temporary variables/nested `if`s.

I also wonder if this condition can actually be false? If it can't, we should use `wxCHECK` here instead of `if`.

vadz commented on src/osx/cocoa/textctrl.mm:
```diff
@@ -1033,26 +1137,47 @@ - (void)controlTextDidEndEditing:(NSNotification *)aNotification
 {
-    if (m_textView && position >=0)
-    {
+    if ( m_textView && position >= 0 )
```
I'll fix this myself, but, for the future changes, please avoid making whitespace-only changes together with other significant ones. TIA!

Dan-Korn replied:
Sorry, trying to follow the coding standards.

vadz commented on src/osx/cocoa/textctrl.mm:
```diff
@@ -1142,80 +1367,256 @@ - (void)controlTextDidEndEditing:(NSNotification *)aNotification
-            [dict release];
         }
     }
+    @catch (NSException *exception)
```
Why do we do need to do this? I.e. why should any exceptions happen here and, if they do, why do we just log them without doing anything else?

Dan-Korn replied:
There are subtle cases where the selection range is not valid, or where the font selected can't be applied to the text because it doesn't support the selected characters.  Letting those exceptions out of the Objective-C code is bad.

vadz commented on src/osx/cocoa/textctrl.mm:
```diff
@@ -1142,80 +1367,256 @@ - (void)controlTextDidEndEditing:(NSNotification *)aNotification
     if (m_textView)
+    {
         [m_textView setContinuousSpellCheckingEnabled: check];
+        [m_textView setAutomaticSpellingCorrectionEnabled: check];
```
Hmm, I wonder why hadn't this been done when adding this code initially. Can you think of any reason why we would _not_ want to do this?

csomor replied:
because autocorrection quite frequently can be false, at least for me when writing in different languages or dialects, that way you just get a hint, if you see the spellchecker is right you can let it correct, but otherwise if it's done automatically I end up writing the same correct word several times, fighting the spell checker

vadz commented on src/osx/cocoa/textctrl.mm:
```diff
@@ -1114,19 +1325,33 @@ - (void)controlTextDidEndEditing:(NSNotification *)aNotification
+            // It's not obvious whether strikethrough is denoted
+            // definitively by wxTEXT_ATTR_EFFECT_STRIKETHROUGH
+            // or by wxTEXT_ATTR_FONT_STRIKETHROUGH, so handle both.
+            int strikeThroughStyle = NSUnderlineStyleNone; // should be type NSUnderlineStyle
```
Why isn't it declared as being of the proper type?

Dan-Korn replied:
See discussion on wx-dev.  Declaring it as NSUnderlineStyle directly fails the xcode6.4 build test.

vadz reviewed (commented):
I haven't actually reviewed the biggest change, i.e. the code setting the attributes, but I guess you've tested it already, so let's just assume that it works as intended.

And while there are a few other remarks/questions, globally this looks pretty good and parts of this could be applied right now. Other parts would need to wait for documentation etc, as discussed on the mailing list.

Thanks again for your contribution!

Dan-Korn committed:
Fix more build errors for 'Compiler: clang Xcode: xcode6.4 C++'

Dan-Korn committed:
Add doc for wxTextEntry methods EmptyUndoBuffer, GetScaleFactor, and SetScaleFactor

Dan-Korn committed:
Merge branch 'osx-style-fixes' of https://github.com/Dan-Korn/wxWidgets into osx-style-fixes

Dan-Korn commented:
The message of the exception being thrown is
> [NSBigMutableString substringWithRange:]: Range {0, 3} out of bounds; string length 0
> which doesn't help me at all to understand what's going wrong.
> Any ideas?

Well, something with a length of zero, but I'd have to see how that was being set up.  How can I reproduce this?

vadz commented:
> How can I reproduce this?

You just need to run the full test, i.e.

    ./test_gui TextCtrlTestCase

TIA for looking at this!

Dan-Korn commented:
> You just need to run the full test, i.e.
> ./test_gui TextCtrlTestCase

I did that against my complete change set, and I got a bunch of errors, including this:
-------------------------------------------------------------------------------
TextCtrlTestCase
  UndoRedo
-------------------------------------------------------------------------------
../../tests/controls/textctrltest.cpp:74
...............................................................................

../../tests/controls/textctrltest.cpp:74: FAILED:
due to unexpected exception with message:
  Unknown exception

So a few questions here:

1. How are you seeing the exception message and call stack, when all I see is "Unknown exception"? Is there a way I can run this under a debugger, such as Xcode?
2. Why doesn't that test get run when I do "./test_gui -c UndoRedo TextCtrlTestCase"?
3. How is the wxTextCtrl being created?  With which flags?

vadz commented:
Of course, you can debug it. I use lldb, you can use Xcode if you prefer.

The reason it doesn't work with just `-c UndoRedo` is long and not very interesting (I could explain it later when I'm not on mobile), but basically this test just doesn't run at all in this case.

You can look at the source file to see how the control is being created, but there is nothing special about it AFAICS.

vadz commented:
It looks like we're stuck with this PR. I don't have neither time nor knowledge needed to debug it and while I might still try to do it before 3.1.3, I really can't guarantee that this happens. And we can't integrate code which fails the simplest possible unit test...

csomor committed:
fixing the attribute name for underline color

csomor commented:
the reason for the crash is the fact that all the controls on a tlw usually share one common NSUndoManager, so the case is never truly set up anew. I've experimented with a private undo manager per control, then the exception does not occur anymore.

Dan-Korn commented:
> the reason for the crash is the fact that all the controls on a tlw usually share one common NSUndoManager, so the case is never truly set up anew. I've experimented with a private undo manager per control, then the exception does not occur anymore.

Thanks for looking at this again.  I admittedly found the relationship between the NSUndoManager and the NSTextView confusing.  And the way it works on this PR is definitely not perfect, though IMO it's better than no undo-redo at all.  (Even if it fails some of the automated tests, because isn't doing nothing at all also a failure?)

Anyway, if there's anything I can do to help move along this PR, please let me know.

csomor commented:
You are very welcome, thank you for your effort. Since we now know why it crashes - (and if it has its own undo-manager it passes the tests, so your implementation works correctly :-) and this is more than before ...

I'm experimenting how NSUndoManager could be integrated even more, on a general level, eg adjusting the menu items, like it does if I allow the selectors undo: and redo: in menuitem.mm, but we have to support our own commands, so this is not the right solution, we cannot use NSUndoManager everywhere, but it helps for trying to find out what should be possible ...

I definitely want to add this before 3.1.5 so that we can get things tested before 3.2 release. Maybe I'll apply it in two parts. I'll come back if I need further help or explanations.

vadz commented:
@csomor Do you still intend to come back to this? There is clearly a lot of useful code here, but it also still needs quite some work (and tests...) to be integrated. I would prefer, of course, if you could do it, but if you won't (before 3.1.5), then I'll make another effort to do it myself, so please let me know. TIA!

csomor commented:
@vadz Ok, I found everything, I don't want to add the private undo manager for the 'normal' apps, there it makes sense to have everything as it usually is, ie sharing the tlw's undo manager. But in the test case I'd trigger my 'private' undo-manager code. To we have some kind of environment variable or a switch that I can consult to find out inside the textctrl.mm whether I'm running tests ?

vadz commented:
Sorry, I don't quite understand what prevents the tests from passing when using the normal case, i.e. the shared undo manager? It would seem to be preferable to test the version of the code we actually use.

I don't think we have any way of knowing whether we're running the test suite right now, but it could be added if really needed, of course. The simplest is probably to just define `WX_RUNNING_TESTS` environment variable.

csomor commented:
> Sorry, I don't quite understand what prevents the tests from passing when using the normal case, i.e. the shared undo manager? It would seem to be preferable to test the version of the code we actually use.

sure, but the current setup of recreating controls on the same tlw seems to confound the undo manager which is shared, I don't know whether it is the fact that no true event loop is cleaning up things. So I just wanted to have this at least run through tests again.

vadz commented:
Could we just recreate a separate TLW for these tests to fix this? If so, this could be the simplest solution.

csomor commented:
> Could we just recreate a separate TLW for these tests to fix this? If so, this could be the simplest solution.

yes, I'm not sure why it is working with an older action of three characters at all, whether ids just map with a previously created control. I'm doing some tests now.

csomor commented:
ok

putting a 

`[m_undoManager removeAllActions];`

when creating the control helps with the crashes. Now of course we'd need some proper naming to call this from the outside. It cannot stay there. Then no private undo manager is needed. 

What I'm testing next: As on OSX Undo-Managers are shared at tlw-level, it could also be a function there or completely standalone, taking a NSWindow* as param.

csomor commented:
ok, the latter also works, along the lines:

```
void wxOSXRemoveAllUndoActions( WXWindow tlw)
{
    [[tlw undoManager] removeAllActions];
}
```

and then before running the test
```
wxFrame* topFrame = wxDynamicCast(wxTheApp->GetTopWindow(), wxFrame);
wxOSXRemoveAllUndoActions(topFrame->GetWXWindow());

```

Dan-Korn committed:
Remove methods zoomToActualSize, zoomIn, zoomOut.

From @interface wxNSTextScrollView.

Dan-Korn committed:
Handle fonts by full, family, style, or PS name.

Add enum wxTextAttrFontNameType, and wxTextAttr
member m_FontNameType and Get/Set methods.
The default is wxTEXT_ATTR_FONTNAME_TYPE_FAMILY,
to specify the font family name.  Other options are:
wxTEXT_ATTR_FONTNAME_TYPE_FULL, for a system-
or language-dependent name;
wxTEXT_ATTR_FONTNAME_TYPE_STYLE, for a
style (sub-family) name, such as "Bold"; and
wxTEXT_ATTR_FONTNAME_TYPE_POSTSCRIPT,
which is an unambiguous way to select a particular
font style.
Implement wxTEXT_ATTR_FONTNAME_TYPE_POSTSCRIPT
for Cocoa (macOS) in wxNSTextViewControl::SetStyle().

vadz commented:
Thanks for returning to this, but could you please explain what is the status of this PR now? Is it supposed to work, i.e. should the tests pass now or do they still fail?

TIA!

Dan-Korn commented:
Hi @vadz, I'm not sure what tests pass.  I was just checking in some changes I had locally that I didn't want to be lost.

GitHub says all checks have failed due to conflicts, but it won't let me see them.  The "Resolve conflicts" button is grayed out, and it says "Only those with write access to this repository can merge pull requests."  What can be done to resolve the conflicts?  What can I do to help?

vadz commented:
I think the best/simplest for you might be to just create a new PR.

If you want to keep working with this one, you need to either rebase it on or merge it with the latest master, resolving conflicts locally, and then push again. If you have troubles doing it, I can do it for you but you need to check the "Enable changes from maintainers" checkbox (or something like that) to give me permission to push into your existing branch and be prepared to update it after a force push.

vadz committed:
Merge branch 'master' into osx-style-fixes

vadz commented:
I've merged master into this branch to resolve the conflicts (just a single one, in fact). I'm still not really qualified to review tons of new code in this PR... The parts related to undo/redo seem relatively straightforward to me, however, so if they pass the tests now, I'd be glad to cherry pick those.

For the style and scale factor changes, I am much less sure. Do they have any chance of being implemented under the other platforms in the future?

vadz commented:
Hmm, merge went without conflicts, but resulted in wrong code because it didn't remove the use of `To{Float,Int}PointSize()` from everywhere.

vadz commented:
I think this is never going to be cleaned and applied by now, so it's probably useless to keep it open.
